x-pack/platform/plugins/shared/cases/server/services/incremental_id/index.ts

```diff
@@ -0,0 +1,391 @@
+      namespaces: [namespace],
+    });
+  }
+
+  /**
+   * Gets or creates the case id incrementer SO for the given namespace
+   * @param namespace The namespace of the case id incrementor so
+   */
+  async getOrCreateCaseIdIncrementerSo(
+    namespace: string
+  ): Promise<SavedObject<CaseIdIncrementerPersistedAttributes>> {
+    try {
+      const [latestAppliedId, incrementerResponse] = await Promise.all([
+        // Get the latest applied id by looking at the case saved objects
+        await this.getLastAppliedIdForSpace(namespace),
+        // Get the case id incrementer saved object
+        await this.getCaseIdIncrementerSo(namespace),
+      ]);
+      this.logger.debug(`Latest applied ID to a case for ${namespace}: ${latestAppliedId}`);
+
+      const actualLatestId = latestAppliedId || 0;
+
+      const incrementerSO = incrementerResponse?.saved_objects[0];
+
+      // We should not have multiple incrementer SO's per namespace, but if we do, let's resolve that
+      if (incrementerResponse.total > 1) {
+        this.logger.error(
+          `Only 1 incrementer should exist, but multiple incrementers found in ${namespace}. Resolving to max incrementer.`
+        );
+        return this.resolveMultipleIncrementerSO(
+          incrementerResponse.saved_objects,
+          actualLatestId,
+          namespace
+        );
+      }
+
+      // Only one incrementer SO exists
+      if (incrementerResponse.total === 1) {
+        // If we have matching incremental ids, we're good
+        const idsMatch = actualLatestId === incrementerSO.attributes.last_id;
+        if (idsMatch || incrementerSO.attributes.last_id >= actualLatestId) {
+          this.logger.debug(
+            `Incrementer found for ${namespace} with matching or bigger id. No changes needed.`
+          );
+          return incrementerSO;
+        } else {
+          // Otherwise, we're updating the incrementer SO to the highest value
+          this.logger.debug(
+            `Incrementer found for ${namespace} with id ${incrementerSO.attributes.last_id}. Updating to ${actualLatestId}.`
+          );
+          return this.incrementCounterSO(incrementerSO, actualLatestId, namespace);
+        }
+      } else {
+        // At this point we assume that no incrementer SO exists
+        this.logger.debug(`No incrementer found for ${namespace}. Creating a new one.`);
+        return this.createCaseIdIncrementerSo(namespace);
+      }
+    } catch (error) {
+      throw new Error(`Unable to use an existing incrementer: ${error}`);
+    }
+  }
+
+  /**
+   * Resolves the situation when multiple incrementer SOs exists
+   */
+  public async resolveMultipleIncrementerSO(
+    incrementerQueryResponse: Array<SavedObjectsFindResult<CaseIdIncrementerPersistedAttributes>>,
+    latestAppliedId: number,
+    namespace: string
+  ) {
+    // Find the incrementer with the highest ID
+    const incrementerWithHighestId = incrementerQueryResponse.reduce<
```

Comment:
nit, it may be a bit more performant to just do a single loop, and track these in two variables? I do prefer more implicit code, but I think overall performance is important for all the methods here in any worst case scenarios

`let incrementerWithHighestId = incrementerQueryResponse[0]`
`let incrementersToDelete = []; // Every time a larger incrementer is found, push the smaller one into the array`